src/metabase/driver/druid/query_processor.clj

```diff
@@ -163,56 +163,72 @@
                  "UTC")
    :locale   "en-US"})
 
-(defn- extract:js [& function-str-parts]
+(defn- extract:js
+  "Create an extraction function from JavaScript -- see http://druid.io/docs/0.9.1.1/querying/dimensionspecs.html#javascript-extraction-function"
+  [& function-str-parts]
   {:pre [(every? string? function-str-parts)]}
   {:type     :javascript
    :function (s/replace (apply str function-str-parts) #"\s+" " ")})
 
-(defn- unit->extractionFn
-  "JODA date format strings for each datetime unit. [Described here.](http://www.joda.org/joda-time/apidocs/org/joda/time/format/DateTimeFormat.html)."
-  [unit]
-  (case unit
-    :default         (extract:timeFormat "yyyy-MM-dd'T'HH:mm:ssZ")
-    :minute          (extract:timeFormat "yyyy-MM-dd'T'HH:mm:00Z")
-    :minute-of-hour  (extract:timeFormat "mm")
-    :hour            (extract:timeFormat "yyyy-MM-dd'T'HH:00:00Z")
-    :hour-of-day     (extract:timeFormat "HH")
-    :day             (extract:timeFormat "yyyy-MM-ddZ")
-    :day-of-week     (extract:js "function (timestamp) {"
-                                 "  var date = new Date(timestamp);"
-                                 "  return date.getDay() + 1;"
-                                 "}")
-    :day-of-month    (extract:timeFormat "dd")
-    :day-of-year     (extract:timeFormat "DDD")
-    :week            (extract:js "function (timestamp) {"
-                                 "  var date     = new Date(timestamp);"
-                                 "  var firstDOW = new Date(date - (date.getDay() * 86400000));"
-                                 "  var month    = firstDOW.getMonth() + 1;"
-                                 "  var day      = firstDOW.getDate();"
-                                 "  return '' + firstDOW.getFullYear() + '-' + (month < 10 ? '0' : '') + month + '-' + (day < 10 ? '0' : '') + day;"
-                                 "}")
-    :week-of-year    (extract:timeFormat "ww")
-    :month           (extract:timeFormat "yyyy-MM-01")
-    :month-of-year   (extract:timeFormat "MM")
-    :quarter         (extract:js "function (timestamp) {"
-                                 "  var date         = new Date(timestamp);"
-                                 "  var month        = date.getMonth() + 1;" ; js months are 0 - 11
-                                 "  var quarterMonth = month - ((month - 1) % 3);"
-                                 "  return '' + date.getFullYear() + '-' + (quarterMonth < 10 ? '0' : '') + quarterMonth + '-01';"
-                                 "}")
-    :quarter-of-year (extract:js "function (timestamp) {"
-                                 "  var date = new Date(timestamp);"
-                                 "  return Math.floor((date.getMonth() + 3) / 3);"
-                                 "}")
-    :year            (extract:timeFormat "yyyy")))
+(def ^:private ^:const unit->extraction-fn
+  {:default         (extract:timeFormat "yyyy-MM-dd'T'HH:mm:ssZ")
```

Comment:
In Clojure maps can act as functions, e.g.

```clojure
(def ^:private ^:const my-map {:a 1, :b 2})

(my-map :a) -> 1
```

So the difference with making something a constant map vs using a function with `case` would be that the entire map will be built a single time during compile time and using it is a simple key-value lookup. For a function call there's the key lookup in the `:case` statement and then the resulting value is constructed at runtime